File: tests/schema/resource_links.py
"""Pure typed links between descriptor catalogs and locked resource declarations."""
import argparse
import json
from pathlib import Path
import re

import bindings as b
import resource_lock as r
import topology as t

MODEL_TOKEN = re.compile(r"[A-Za-z_][A-Za-z0-9_]{0,63}\Z")


def validate(document):
    t.nesting(document)
    t.fields(document, "format version topology lock assets symbols models", "$")
    t.require(document["format"] == "simnodus-resource-links" and document["version"] == "0.1",
              "version", "$", "Expected resource links 0.1")
    topology = document["topology"]
    stats = b.validate(topology)
    inventory = r.validate(document["lock"])
    files = {(dependency["id"], file["id"])
             for dependency in document["lock"]["dependencies"] for file in dependency["files"]}
    assets, targets = {}, set()
    values = t.array(document["assets"], "$/assets")
    t.require(len(values) <= r.MAX_FILES, "budget", "$/assets", "Asset budget exceeded")
    for index, value in enumerate(values):
        loc = f"$/assets/{index}"
        t.fields(value, "id kind dependency resource", loc)
        for field in ("id", "dependency", "resource"):
            t.identifier(value[field], loc + "/" + field)
        t.require(type(value["kind"]) is str and value["kind"] in ("symbol-svg", "model-spice"),
                  "kind", loc, "Unsupported declared asset kind")
        t.require(value["id"] not in assets, "id", loc, "Duplicate asset ID")
        target = (value["dependency"], value["resource"])
        t.require(target in files, "reference", loc, "Unknown locked dependency/resource pair")
        t.require(target not in targets, "reference", loc, "One locked resource has one declared role")
        targets.add(target)
        assets[value["id"]] = value
    linked, model_entries = 0, 0
    for collection, kind in (("symbols", "symbol-svg"), ("models", "model-spice")):
        loc = "$/" + collection
        mapping = document[collection]
        expected = {descriptor["id"] for descriptor in topology[collection]}
        t.require(type(mapping) is dict and set(mapping) == expected, "mapping", loc,
                  "Every descriptor requires an explicit resource binding or null")
        for descriptor, target in mapping.items():
            if target is None:
                continue
            if collection == "models":
                where = loc + "/" + descriptor
                t.fields(target, "asset entrypoint terminal_map parameter_map", where)
                entrypoint = target["entrypoint"]
                t.require(type(entrypoint) is str and MODEL_TOKEN.fullmatch(entrypoint) is not None,
                          "mapping", where, "Invalid declared model entrypoint")
                definition = next(value for value in topology["models"] if value["id"] == descriptor)
                for key, source in (("terminal_map", "terminals"), ("parameter_map", "parameters")):
                    bindings = target[key]
                    t.require(type(bindings) is dict and set(bindings) == {v["id"] for v in definition[source]},
                              "mapping", where, "Model source mapping must cover the interface")
                    for token in bindings.values():
                        t.require(type(token) is str and MODEL_TOKEN.fullmatch(token) is not None,
                                  "mapping", where, "Invalid declared source token")
                    t.require(len({v.lower() for v in bindings.values()}) == len(bindings),
                              "mapping", where, "Aliased model source tokens")
                    model_entries += len(bindings)
                model_entries += 1
                target = target["asset"]
            t.identifier(target, loc + "/" + descriptor)
            t.require(target in assets, "reference", loc, "Unknown asset ID")
            t.require(assets[target]["kind"] == kind, "kind", loc, "Descriptor and asset kinds differ")
            linked += 1
    t.require(stats["declared_entities"] + stats["parameter_entries"] + stats["descriptor_entries"]
              + len(assets) + len(document["symbols"]) + len(document["models"]) + model_entries <= t.MAX_ENTITIES,
              "budget", "$", "Combined descriptor/resource binding budget exceeded")
    return dict(descriptors_linked=linked, assets=len(assets), **inventory,
                resource_interfaces_verified=False, topology_stats=stats)


def parse(raw):
    t.require(len(raw) <= t.MAX_BYTES, "input", "$", "Input byte budget exceeded")

    def pairs(items):
        result = {}
        for key, value in items:
            t.require(key not in result, "input", "$", "Duplicate JSON key")
            result[key] = value
        return result

    def constant(_):
        raise t.Invalid("input", "$", "Non-finite JSON literal")

    try:
        document = json.loads(raw.decode("utf-8"), object_pairs_hook=pairs, parse_constant=constant)
        validate(document)
        return document
    except (UnicodeError, ValueError, RecursionError) as error:
        if isinstance(error, t.Invalid):
            raise
        raise t.Invalid("input", "$", "Invalid encoding, JSON or nesting") from error


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("document", type=Path)
    args = parser.parse_args()
    try:
        with args.document.open("rb") as source:
            document = parse(source.read(t.MAX_BYTES + 1))
        print(json.dumps(dict(status="valid-resource-links-only", **validate(document))))
        return 0
    except (t.Invalid, OSError) as error:
        diagnostic = error.diagnostic() if isinstance(error, t.Invalid) else dict(code="input", location="$", message="Cannot read input")
        print(json.dumps(dict(status="invalid", **diagnostic)))
        return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: src/application/resource_links.cpp
// Copyright (c) 2026 Ricardo Kerschbaumer
// SPDX-License-Identifier: MIT
#include "application/resource_links.hpp"
#include "application/captured_validation_internal.hpp"
#include <algorithm>
#include <map>
#include <new>
#include <set>

namespace simnodus {
namespace {
using Index = std::size_t;
using Fields = std::map<std::string, Index>;
struct Asset { std::string kind; };
class Validator {
public:
    explicit Validator(std::shared_ptr<const DeclarationSyntax> source) : source_(std::move(source)) {}
    ResourceLinkDeclaration run()
    {
        const auto root = exact(0, {"format", "version", "topology", "lock", "assets", "symbols", "models"});
        require(text(root.at("format"), "version") == "simnodus-resource-links"
            && text(root.at("version"), "version") == "0.1", "version", 0);
        auto topology = detail::captured_bindings(source_, root.at("topology"));
        auto lock = detail::captured_lock(source_, root.at("lock"));
        std::set<std::pair<std::string, std::string>> files, targets;
        for(const auto& row : lock.requests) files.emplace(row.dependency, row.resource);
        const auto values = array(root.at("assets"));
        require(values.size() <= resource_max_files, "budget", root.at("assets"));
        std::map<std::string, Asset> assets;
        for(const auto token : values) {
            const auto value = exact(token, {"id", "kind", "dependency", "resource"});
            const auto id = identifier(value.at("id"));
            const auto dependency = identifier(value.at("dependency"));
            const auto resource = identifier(value.at("resource"));
            const auto kind = text(value.at("kind"), "kind");
            require(kind == "symbol-svg" || kind == "model-spice", "kind", value.at("kind"));
            require(assets.emplace(id, Asset{kind}).second, "id", value.at("id"));
            const auto target = std::make_pair(dependency, resource);
            require(files.contains(target) && targets.insert(target).second, "reference", token);
        }
        const auto topology_fields = object(root.at("topology"));
        std::size_t linked = 0, extra = assets.size();
        for(const bool model : {false, true}) {
            const auto collection = model ? "models" : "symbols";
            Fields descriptors;
            for(const auto token : array(topology_fields.at(collection)))
                descriptors.emplace(text(object(token).at("id"), "id"), token);
            const auto mappings = complete_map(root.at(collection), descriptors);
            extra += mappings.size();
            for(const auto& [id, target] : mappings) {
                if(source_->tokens[target].kind == JsonKind::null) continue;
                auto asset_token = target;
                if(model) {
                    const auto fields = exact(target, {"asset", "entrypoint", "terminal_map", "parameter_map"});
                    model_token(fields.at("entrypoint"));
                    const auto descriptor = object(descriptors.at(id));
                    for(const bool parameter : {false, true}) {
                        Fields interface;
                        for(const auto token : array(descriptor.at(parameter ? "parameters" : "terminals")))
                            interface.emplace(text(object(token).at("id"), "id"), token);
                        const auto map_index = fields.at(parameter ? "parameter_map" : "terminal_map");
                        const auto mapping = complete_map(map_index, interface);
                        std::set<std::string> names;
                        for(const auto& [source, destination] : mapping) {
                            (void)source;
                            auto name = model_token(destination);
                            for(auto& c : name) if(c >= 'A' && c <= 'Z') c = static_cast<char>(c + ('a' - 'A'));
                            require(names.insert(name).second, "mapping", destination);
                        }
                        extra += mapping.size();
                    }
                    ++extra;
                    asset_token = fields.at("asset");
                }
                const auto asset = identifier(asset_token);
                const auto found = assets.find(asset);
                require(found != assets.end(), "reference", asset_token);
                require(found->second.kind == (model ? "model-spice" : "symbol-svg"), "kind", asset_token);
                ++linked;
            }
        }
        const auto& p = topology.parameters;
        require(p.topology.declared_entities + p.parameter_entries + topology.descriptor_entries + extra <= 4096, "budget", 0);
        return {std::move(topology), std::move(lock), linked, assets.size(), false};
    }
private:
    std::shared_ptr<const DeclarationSyntax> source_;
    void require(bool value, const char* code, Index index) const
    {
        if(!value) throw LockError{code, source_->tokens[index].begin};
    }
    Fields object(Index index, const char* code = "shape") const
    {
        const auto& token = source_->tokens[index];
        require(token.kind == JsonKind::object, code, index);
        Fields result;
        for(auto child = index + 1; child < token.next; child = source_->tokens[child + 1].next)
            result.emplace(source_->tokens[child].decoded, child + 1);
        return result;
    }
    Fields exact(Index index, std::initializer_list<const char*> names) const
    {
        auto result = object(index);
        require(result.size() == names.size(), "shape", index);
        for(const auto name : names) require(result.contains(name), "shape", index);
        return result;
    }
    Fields complete_map(Index index, const Fields& expected) const
    {
        auto result = object(index, "mapping");
        require(result.size() == expected.size(), "mapping", index);
        for(const auto& [id, value] : result) { (void)value; require(expected.contains(id), "mapping", index); }
        return result;
    }
    std::vector<Index> array(Index index) const
    {
        const auto& token = source_->tokens[index];
        require(token.kind == JsonKind::array, "shape", index);
        std::vector<Index> result;
        for(auto child = index + 1; child < token.next; child = source_->tokens[child].next) result.push_back(child);
        return result;
    }
    std::string text(Index index, const char* code) const
    {
        require(source_->tokens[index].kind == JsonKind::string, code, index);
        return source_->tokens[index].decoded;
    }
    std::string identifier(Index index) const
    {
        const auto value = text(index, "id");
        require(!value.empty() && value.size() <= 64 && value[0] >= 'a' && value[0] <= 'z'
            && std::all_of(value.begin(), value.end(), [](char c) { return (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9') || c == '_' || c == '-'; }), "id", index);
        return value;
    }
    std::string model_token(Index index) const
    {
        const auto value = text(index, "mapping");
        const auto letter = [](char c) { return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_'; };
        require(!value.empty() && value.size() <= 64 && letter(value[0])
            && std::all_of(value.begin(), value.end(), [&](char c) { return letter(c) || (c >= '0' && c <= '9'); }), "mapping", index);
        return value;
    }
};
}
ResourceLinkResult validate_resource_links(std::string_view bytes)
{
    try {
        const auto input = capture_declaration_syntax(bytes);
        if(const auto* error = std::get_if<IngressError>(&input))
            return LockError{error->code == IngressErrorCode::memory ? "memory" : "input", error->offset};
        return std::make_shared<const ResourceLinkDeclaration>(Validator(std::get<0>(input)).run());
    } catch(const TopologyError& error) { return LockError{topology_error_name(error.code), error.offset}; }
    catch(const ParameterError& error) { return LockError{error.code, error.offset}; }
    catch(const LockError& error) { return error; }
    catch(const std::bad_alloc&) { return LockError{"memory", 0}; }
}
}
